src/main/core/worker.rs

```diff
@@ -11,25 +11,87 @@ use crate::host::process::Process;
 use crate::host::process::ProcessId;
 use crate::host::thread::ThreadId;
 use crate::host::thread::{CThread, Thread};
+use crate::network::network_graph::{IpAssignment, RoutingInfo};
 use crate::utility::counter::Counter;
 use crate::utility::notnull::*;
+use crate::utility::SyncSendPointer;
 
 use std::cell::{Cell, RefCell};
+use std::collections::HashMap;
 use std::sync::atomic::AtomicBool;
 use std::sync::Arc;
 use std::sync::Mutex;
 
 static USE_OBJECT_COUNTERS: AtomicBool = AtomicBool::new(false);
 
-// counters to be used when there is no worker active
+// global counters to be used when there is no worker active
 static ALLOC_COUNTER: Lazy<Mutex<Counter>> = Lazy::new(|| Mutex::new(Counter::new()));
 static DEALLOC_COUNTER: Lazy<Mutex<Counter>> = Lazy::new(|| Mutex::new(Counter::new()));
 static SYSCALL_COUNTER: Lazy<Mutex<Counter>> = Lazy::new(|| Mutex::new(Counter::new()));
 
+// thread-local global state
 std::thread_local! {
     // Initialized when the worker thread starts running. No shared ownership
     // or access from outside of the current thread.
-    static WORKER: OnceCell<RefCell<Worker>> = OnceCell::new();
+    static WORKER: once_cell::unsync::OnceCell<RefCell<Worker>> = once_cell::unsync::OnceCell::new();
+}
+
+// shared global state
+pub static WORKER_SHARED: once_cell::sync::OnceCell<WorkerShared> =
+    once_cell::sync::OnceCell::new();
+
+#[derive(Debug)]
+pub struct WorkerShared {
+    pub ip_assignment: IpAssignment<u32>,
+    pub routing_info: RoutingInfo<u32>,
+    pub host_bandwidths: HashMap<std::net::IpAddr, Bandwidth>,
+    pub dns: SyncSendPointer<cshadow::DNS>,
+}
+
+impl WorkerShared {
+    pub unsafe fn dns(&self) -> *mut cshadow::DNS {
```

Comment:
I think this isn't actually unsafe? If it is unsafe, would be good to document the preconditions for calling this function safely with a `SAFETY` comment. https://rust-lang.github.io/rust-clippy/master/#missing_safety_doc

Reply:
You're right; removed the unsafe.